Use the MHD occupation probability formalism for H I bound-free and bound-bound absorption

ajwheeler opened the pull request:
This has been a long time coming.  I'm going to recheck my equations, a few more plots, etc before considering it good to merge.

solves (hopefully) #149 

Changes
- Uses the MHD theory to adjust occupation fractions for H energy levels for H I bf and bb absorption.  This turns the balmer jump into a soft transition in many stars, which is the motivating reason for PR.
- Eliminates the Karzas & Latter (1961) via Gingerich (1969) via Kurucz (1970) hydrogenic bf opacs.  Instead...
- [Nahar 2021](https://ui.adsabs.harvard.edu/abs/2021Atoms...9...73N) (NORAD) is used for H I bf
- The TOPbase opacs are used for He II.  (This was actually being "double-counted" before, but it really doesn't matter because it's subdominant by many OOM.)
- The simple analytic cross-sections from Kurucz (But several other places as well) are used for bf absorption from n>6.

TODO
- [x] check that we reproduce turbospectrum's balmer series in metal-poor stars.
- [x] Compare to turbospectrum balmer break in sun
- [x] (probably) interpolate the cross-sections more intelligently 
- [ ] clean up commit history
- [x] Probably tweak how things are distributed between files.

![image](https://user-images.githubusercontent.com/711963/219147108-b4d92fcc-3ef5-4277-8982-1874c216496e.png)![image](https://user-images.githubusercontent.com/711963/219147124-fdd12e5c-4e37-4b0f-8d7e-475a6791b812.png)
(This is the sun.  Other stars are much worse.)

segasai commented:
Really cool! I'd be happy to give it for a spin on some hotter stars that I have.

ajwheeler commented:
✅ This produces results similar Turbospectrum/HBOP's.  (colors/line styles are as in the paper, so blue is Korg, orange is turbospec.)

![image](https://user-images.githubusercontent.com/711963/219521504-1abb9396-03f8-493a-8405-2ba7434c877e.png)![image](https://user-images.githubusercontent.com/711963/219521487-371413b1-93cd-4a30-85db-c86cc1b85196.png)

✅ and with the fix I'm about to push, I reproduce Turbospectrum's missing blamer lines in HD122563.
![image](https://user-images.githubusercontent.com/711963/219530378-445726bb-7fcb-472d-bcd3-b9035ce95a52.png)
There's a flag to disable MHD for H lines (not the H I bf absorption), because, as we discussed in the paper, this clashes with observations. MHD is on by default, though.

ajwheeler commented:
The failing test has to do with reference data that needs to be updated.  I will fix in when I'm more completely confident that my implementation is correct.

Surprisingly (to me), my new tolerance ratcheting flagged that the Nahar H I bf cross-sections are *incredibly* close to the reference ones (`atol=1e-20` in instead of  the old `atol=0.007`), which @mabruzzo implemented from Gray's textbook. This warrants further investigation, but probably doesn't need to hold up this PR.  It would probably be an sped improvement to use the gray implementation for synthesis and the Nahar one for testing.

ajwheeler commented:
I've switched over to extrapolating the cross-section by assuming they go like E^3, rather than simple linear interpolation.  The MHD occupation probabilities seem to change so fast as a function of energy/wavelength that it doesn't really matter, but this is probably more physically motivated.

The PR is now totally ready for review.

ajwheeler commented on src/ContinuumAbsorption/ContinuumAbsorption.jl:
```diff
@@ -1,26 +1,18 @@
 include("absorption_He.jl")
-
 include("absorption_ff_positive_ion.jl")
-#used for metal bf absorption
```
I moved these imports to `absorption_metals_bf.jl`, as there's no overhead to duplicate `using`s and it means that reading the code requires a lot less flipping between files.

I realize that it's not consistent with the imports from Korg above, but I think it's slightly better.

mabruzzo reviewed (changes requested):
I have a few minor comments (hopefully they aren't too annoying - they mostly relate to documentation). But otherwise, this looks great!

ajwheeler commented on data/bf_cross-sections/Manifest.toml:
```diff
@@ -0,0 +1,544 @@
+# This file is machine-generated - editing it directly is not advised
+
```
Yeah, it records the exact versions for the sake of reproducibility.  I will add a note to the devdocs page.

ajwheeler commented on src/ContinuumAbsorption/absorption_H.jl:
```diff
@@ -1,44 +1,151 @@
+                frac = 1 - w_upper/w_lower
+                if taper 
+                    # taper of the cross-section past a certain  wavelength redward of the jump, as 
+                    # is done in HBOP. (Not ennabled in Korg calls to this function.)
```
Fixed spelling.  I don't want to say "by default" because there isn't a way to switch it on at all.  It's just left in in case we want it later.

ajwheeler commented on src/ContinuumAbsorption/absorption_H.jl:
```diff
@@ -1,44 +1,151 @@
+                # w_upper/w[n] is the prob that the upper level is dissolved given that the lower isn't
+                frac = 1 - w_upper/w_lower
+                if taper 
+                    # taper of the cross-section past a certain  wavelength redward of the jump, as 
```
I'm referring to the break/classical ionization energy.  I've clarified.

ajwheeler commented on src/ContinuumAbsorption/absorption_H.jl:
```diff
@@ -1,44 +1,151 @@
+                # don't use MHD for the Lyman series limit since it leads to inflated cross-sections
+                # far red of the limit
+                0.0 
+            else  # account for bf absorption redward of the limit because of level dissolution
```
sorry, that's sloppy writing.  I'm referring to the jump/break corresponding to the ionization energy for each n.  I've changed the comment to call it the jump/break.

ajwheeler commented on src/synthesize.jl:
```diff
@@ -91,9 +94,10 @@ function synthesize(atm::ModelAtmosphere, linelist, A_X, λ_start, λ_stop, λ_s
+                    bezier_radiative_transfer=false, ionization_energies=ionization_energies, 
+                    partition_funcs=default_partition_funcs, 
                     log_equilibrium_constants=default_log_equilibrium_constants)
     #work in cm
```
Why would that be a problem?

ajwheeler replied:
Oops, I see.

ajwheeler commented on src/synthesize.jl:
```diff
@@ -50,6 +50,9 @@ solution = synthesize(atm, linelist, A_X, 5000, 5100)
   calculated.
 - `hydrogen_lines` (default: `true`): whether or not to include H lines in the synthesis.
+- `use_MHD_for_hydrogen_lines` (default: `true`): whether or not to use the MHD occupation 
+   probability formalism for hydrogen lines.  (It is always used for hydrogen bound-free absorption.)
+   hydrogen lines.  This is only valid for spherical geometry.
```
The note about spherical geometry is nonsense, I think it's some kind of copy-paste typo.  I've removed and reworded the parenthetical.

mabruzzo commented on src/statmech.jl:
```diff
@@ -175,17 +175,10 @@ function chemical_equilibrium_equations(T, nₜ, nₑ, absolute_abundances, ioni
-###################################################################################
 
 """
     hummer_mihalas_w(T, n, E_level, nH, nHe, ne)
```
I think that sounds like a sensible plan! 

More generally, it's a little unclear to me if it would ever stop being a "derived work" even if you rewrite it (since your from-scratch implementation might still be influenced by the earlier work), but in this context I definitely don't think that would be an issue.

ajwheeler replied:
done.

ajwheeler commented:
Thanks, Matthew.  Your comments made me notice that I had made an error which switching to extrapolating the cross-sections by assuming they go like E^3, instead of using naive linear interpolation.  I should have been more suspicious when the differences I was seeing were minuscule. Here's what they really look like:
![image](https://user-images.githubusercontent.com/711963/223532528-51d4c6d3-4d88-49bd-b6b6-52e476a63a6a.png)![image](https://user-images.githubusercontent.com/711963/223532558-ebd566cd-1e15-4498-9e71-b5878d373a8f.png)

This gives a slightly larger disagreement with turbospectrum for the balmer jump (contrast blue=Korg and orange=TS).  The difference isn't so big as to indicate that this approach is obviously wrong.
![image](https://user-images.githubusercontent.com/711963/223531630-d329dec1-3e6a-4e3a-81a0-8ca0a2b0c2d6.png)![image](https://user-images.githubusercontent.com/711963/223531661-458d9f0e-b861-43dd-a7d4-356759ea32fa.png)

I'm going to have a think about this, and probably email Paul Barklem.  The difference is just a few lines of code, so it makes sense to continue working on this PR in its present state.

I've addressed the initial mistake and the big switch-inside-a-map efficiency problem in a forthcoming commit.

ajwheeler commented on src/ContinuumAbsorption/absorption_H.jl:
```diff
@@ -1,44 +1,151 @@
+            elseif !use_MHD_for_Lyman && n==1
+                # don't use MHD for the Lyman series limit since it leads to inflated cross-sections
+                # far red of the limit
+                0.0 
```
Done.

ajwheeler commented on src/ContinuumAbsorption/absorption_H.jl:
```diff
@@ -1,44 +1,151 @@
+            break
+        end
+
+        # extrapolate the cross-section to lower energies by assuming proportionality to ν^3
```
This was a mistake, sorry.

ajwheeler commented on test/continuum_absorption.jl:
```diff
@@ -501,52 +501,20 @@ end
-            absorption_coef = Korg.ContinuumAbsorption.H_I_bf(ν_vals, T, nH_I/H_I_partition_val,
-                                                              H_I_ion_energy) / nH_I
-            @test maximum(abs.(absorption_coef - ref_absorption_coef)/absorption_coef) < 0.007
+            absorption_coef = Korg.ContinuumAbsorption.H_I_bf(ν_vals, T, 1, 0, 0, 1/H_I_partition_val)
```
done.

ajwheeler commented on src/ContinuumAbsorption/absorption_H.jl:
```diff
@@ -1,44 +1,151 @@
+
+
+"""
+    H_I_bf(νs, T, nH, nHe, ne, invU_H; n_max_MHD=6, use_hubeny_generalization=false, 
```
Good catch.  Done.

ajwheeler commented on src/ContinuumAbsorption/absorption_H.jl:
```diff
@@ -1,44 +1,151 @@
-           kwargs...)
+function H_I_bf(νs, T, nH, nHe, ne, invU_H; n_max_MHD=6, use_hubeny_generalization=false, 
+                taper=false, use_MHD_for_Lyman=false)
+    χ = ionization_energies[1][1]
```
I'm tentatively against this, because I don't want to check the bounds, and I don't want the kwarg to not do anything.

ajwheeler commented on src/ContinuumAbsorption/absorption_H.jl:
```diff
@@ -1,44 +1,151 @@
+
+By default, detailed cross-sections for n=1 through n=6 are included, with the rest modelled with a 
+simple integral. This can be overridden by setting `n_max_MHD`. TODO
+
```
done.

ajwheeler commented:
I've addressed all but 1 or your comments, and fixed the bug that you turned up.  If you have time, it probably makes sense to take another look at `H_I_bf`, as I've restructured it a decent amount.

ajwheeler commented:
I've edited my comment on E^-3 extrapolation to correct a dumb mistake and reflect the fix I just pushed. (posting to ping people's notifications.)
